mtags/src/main/scala/scala/meta/internal/mtags/JavaToplevelMtags.scala

```diff
@@ -0,0 +1,272 @@
+        case ')' => (Token.RParen, false)
+        case '[' => (Token.LBracket, false)
+        case ']' => (Token.RBracket, false)
+        case '"' => (quotedLiteral('"'), false)
+        case '\'' => (quotedLiteral('\''), false)
+        case '/' =>
+          reader.nextChar()
+          val next = reader.ch
+          next match {
+            case '*' =>
+              skipMultilineComment(prevStar = false)
+              toNextNonWhiteSpace()
+              parseToken
+            case '/' =>
+              skipLine
+              toNextNonWhiteSpace()
+              parseToken
+            case _ =>
+              val token = kwOrIdent(
+                reader.charOffset - 1,
+                new StringBuilder().append(first).append(next)
+              )
+              (token, true)
+          }
+        case _ =>
+          val token = kwOrIdent(reader.charOffset, new StringBuilder(first))
+          (token, true)
+      }
+    }
+
+    toNextNonWhiteSpace()
+    val (t, didNextChar) = parseToken
+    if (!didNextChar) reader.nextChar()
+    t
+  }
+
+  private def readPaths: List[Token.WithPos] = {
+    val builder = List.newBuilder[Token.WithPos]
+    @tailrec
+    def loop(): List[Token.WithPos] = {
+      fetchToken match {
+        case t: Token.WithPos =>
+          builder += t
+          loop()
+        case Token.Dot => loop()
+        case _ =>
+          builder.result()
+      }
+    }
+    loop()
+  }
+
+  private def isWhitespace(ch: Char): Boolean = {
+    ch match {
+      case ' ' | '\t' | CR | LF | FF => true
+      case _ => false
+    }
+  }
+
+  private def skipBody: Unit = {
+    @tailrec
+    def skipToFirstBrace: Unit =
+      fetchToken match {
+        case Token.LBrace | Token.EOF => ()
+        case _ =>
+          skipToFirstBrace
+      }
+
+    @tailrec
+    def skipToRbrace(open: Int): Unit = {
+      fetchToken match {
+        case Token.RBrace if open == 1 => ()
+        case Token.RBrace =>
+          skipToRbrace(open - 1)
+        case Token.LBrace =>
+          skipToRbrace(open + 1)
+        case Token.EOF => ()
+        case _ =>
+          skipToRbrace(open)
+      }
+    }
+
+    skipToFirstBrace
+    skipToRbrace(1)
+  }
+
+  private def skipLine: Unit =
+    while ({ val ch = reader.ch; ch != SU && ch != '\n' }) reader.nextChar()
+
+  @tailrec
+  private def toNextNonWhiteSpace(): Unit = {
+    if (isWhitespace(reader.ch)) {
+      reader.nextChar()
+      toNextNonWhiteSpace()
+    }
+  }
+
+  private def readCurrentLine: String = {
+    def loop(builder: StringBuilder): String = {
+      val ch = reader.ch
+      if (ch == '\n' || ch == SU)
+        builder.mkString
+      else {
+        val next = builder.append(ch)
+        reader.nextChar()
+        loop(next)
+      }
+    }
+
+    val lineOffset = reader.lineStartOffset
+    val existing = input.text.substring(lineOffset, reader.charOffset)
+    loop(new StringBuilder().append(existing))
+  }
+
+}
+
+object JavaToplevelMtags {
```

Comment:
Do you think it would make sense to use the existing Java scanner from eclipse? Or would that be much slower?

Reply:
I didn't know that there is a scanner from eclipse.
I think this impl is quite compact and I would keep it instead investigating if it possible to switch on ecplipse one and relying on it.

Reply:
We use that scanner for EditDistance, but it's fine to leave it at this.

Reply:
> We use that scanner for EditDistance,

🤔   it might fit. The only issue is that it will bring eclipse libraries as dependencies into mtags. 

Yep, lets keep it as is. If there will be some issues then we can try to replace it instead of fixing this code.